graph_algorithms/algorithms/examples/louvain.gsql

```diff
@@ -1,80 +1,295 @@
+                            ACCUM s.@weightToCluster += (t.@cid -> e.weight)
+                            POST-ACCUM INT bestCluster = s.@cid,
+                                       FLOAT maxDeltaQ = 0.0,
+                                       FLOAT deltaQ_new = 0.0,
+                                       FOREACH (cluster, weightToC) IN s.@weightToCluster DO   #would be better if this can be distributed
+                                               FLOAT incident = @@totIncidentCluster.get(cluster),
+                                               deltaQ_new = weightToC - 2 * incident * s.@weight/ @@totalWeight,
+                                               IF deltaQ_new > maxDeltaQ OR (abs(deltaQ_new - maxDeltaQ) < epsilon AND cluster < bestCluster) THEN   # when deltaQ_new is equal to maxDeltaQ, and the cluster label is smaller, also change 
+                                                       maxDeltaQ = deltaQ_new,
+                                                       bestCluster = cluster
+                                               END
+                                       END,
+                                       IF s.@cid != bestCluster THEN 
+                                               @@totIncidentCluster += (s.@cid -> (-1 * s.@weight)),
+                                               @@totIncidentCluster += (bestCluster -> s.@weight),
+                                               s.@cid = bestCluster
+                                       END,
+                                       s.@weightToCluster.clear();
+                        loop = loop + 1;
+                END;
+                last_modularity = @@modularity;
+                @@modularity = 0;
+                T1 = SELECT s
+                     FROM Start:s-(Friend:e)-:t
+                     WHERE s.@cid == t.@cid
+                     ACCUM @@modularity += e.weight - s.@weight * t.@weight / (@@totalWeight);
+                @@modularity = @@modularity / @@totalWeight;                      
+                PRINT iteration AS Phase1Iter, @@modularity;
+                log(debug > 0, "[redrain]#move", iteration, @@modularity);
         END;
 
-        Start = {ANY};
+# Phase 2 --  Merge     
+        iteration2 = 0;
+        WHILE (iteration2 < 2 OR @@modularity2 - last_modularity2 > epsilon) LIMIT iter2 DO
+                iteration2 += 1;
+                Start = SELECT s
+                        FROM Start:s
+                        ACCUM s.@uid = s.@cid;      
+                # Select the vertices with minimal internal id to represent the coarsened graph
+                Start = SELECT s
+                        FROM Start:s 
+                        ACCUM @@representMap += (s.@cid -> s);
+        
+                FOREACH (key, value) IN @@representMap DO
+                        @@representSet += value;                       
+                END;      
+                represent = {@@representSet};
```

Comment:
I think the represent set should be always the initial clusters before the merging. This will make the merging phase exactly the same as the previous phase.

Reply:
In every iteration2, the @cid is changed, so the represent set is also changed.